src/service/repository/repo.rs

```diff
@@ -125,4 +129,173 @@ impl Repository {
             _ => Err(anyhow::anyhow!("No input provided")),
         }
     }
+
+    pub fn get_repo_with_relations(
+        connection: &mut PgConnection,
+        user_id: &Uuid,
+        repo_url: Option<&str>,
+        soft_serve_url: Option<&str>,
+        pagination: &PaginationParams,
+    ) -> Result<PaginatedResponse<RepositoryWithRelations>> {
+        use crate::schema::{challenges, progress, repositories};
+
+        if repo_url.is_some() && soft_serve_url.is_some() {
+            return Err(anyhow::anyhow!(
+                "Multiple parameters are not allowed. Please provide only one parameter."
+            ));
+        }
+
+        let page = pagination.page.unwrap_or(1);
+        let per_page = pagination.per_page.unwrap_or(10);
+        let offset = (page - 1) * per_page;
+
+        let mut query = repositories::table
+            .inner_join(challenges::table)
+            .inner_join(
+                progress::table.on(progress::user_id
+                    .eq(repositories::user_id)
+                    .and(progress::challenge_id.eq(repositories::challenge_id))),
+            )
+            .filter(repositories::user_id.eq(user_id))
+            .into_boxed();
+
+        if let Some(url) = repo_url {
+            query = query.filter(repositories::repo_url.eq(url));
+        }
+
+        if let Some(url) = soft_serve_url {
+            query = query.filter(repositories::soft_serve_url.eq(url));
+        }
+
+        let total: i64 = repositories::table
+            .inner_join(challenges::table)
+            .inner_join(
+                progress::table.on(progress::user_id
+                    .eq(repositories::user_id)
+                    .and(progress::challenge_id.eq(repositories::challenge_id))),
+            )
+            .filter(repositories::user_id.eq(user_id))
+            .count()
+            .get_result(connection)
+            .map_err(|e| {
+                error!("Error getting total count: {}", e);
+                anyhow::anyhow!("Failed to get total count")
+            })?;
+
+        let total_pages = (total as f64 / per_page as f64).ceil() as i64;
```

Comment:
```suggestion
        let mut count_query = repositories::table
            .inner_join(challenges::table)
            .inner_join(
                progress::table.on(progress::user_id
                    .eq(repositories::user_id)
                    .and(progress::challenge_id.eq(repositories::challenge_id))),
            )
            .filter(repositories::user_id.eq(user_id))
            .into_boxed();

        if let Some(url) = repo_url {
            count_query = count_query.filter(repositories::repo_url.eq(url));
        }

        if let Some(url) = soft_serve_url {
            count_query = count_query.filter(repositories::soft_serve_url.eq(url));
        }

        let total: i64 = count_query
            .count()
            .get_result(connection)
            .map_err(|e| {
                error!("Error getting total count: {}", e);
                anyhow::anyhow!("Failed to get total count")
            })?;

        let total_pages = (total as f64 / per_page as f64).ceil() as i64;
```

Reply:
We should apply the same filters to both the actual query and the one which gets the counts I think?  Check this code works as it's just a suggestion.

Reply:
@daviroo I am using the same filter parameters for both the count and the actual query. Your suggestion implies we use the same query for both, which won't work in our case because once `count_query` is used on line 188, we can't reuse it when collecting the results as it'd have been moved. And the query doesn't implement clone; if it did, it'd have been easier to clone and reuse it.